metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala

```diff
@@ -1483,33 +1487,56 @@ class MetalsLanguageServer(
   }
 
   private def autoConnectToBuildServer(): Future[BuildChange] = {
-    for {
-      _ <- disconnectOldBuildServer()
-      maybeBuild <- timed("connected to build server") {
-        if (buildTools.isBloop) bloopServers.newServer(userConfig)
-        else bspServers.newServer()
-      }
-      result <- maybeBuild match {
-        case Some(build) =>
-          connectToNewBuildServer(build)
-        case None =>
-          Future.successful(BuildChange.None)
-      }
-      _ = {
-        treeView.init()
+    def compileAllOpen: BuildChange => Future[BuildChange] = {
+      case change if !change.isFailed =>
+        Future
+          .sequence[Unit, List](
+            compilations
+              .cascadeCompileFiles(buffers.open.toSeq)
+              .ignoreValue ::
+              compilers.load(buffers.open.toSeq) ::
+              Nil
+          )
+          .map(_ => change)
+      case other => Future.successful(other)
+    }
+
+    {
+      for {
+        _ <- disconnectOldBuildServer()
+        maybeBuild <- timed("connected to build server") {
+          if (buildTools.isBloop) bloopServers.newServer(userConfig)
+          else bspServers.newServer()
+        }
+        result <- maybeBuild match {
+          case Some(build) =>
+            val result = connectToNewBuildServer(build)
+            build.onReconnection { reconnected =>
+              connectToNewBuildServer(reconnected)
+                .flatMap(compileAllOpen)
```

Comment:
We also compile, which will effectively just move all the old artifacts to the new BSP directory.